packages/core/bootstrap/src/lib/cache-warmer/reducer.ts

```diff
@@ -26,25 +26,135 @@ export interface SubscriptionData {
+  /**
+   * If a subscription is warming multiple other requests
+   * This will hold a map of the subscription key to the last time it was seen
+   */
+  childLastSeenById?: { [childKey: string]: number }
 }
 
 export interface SubscriptionState {
   [requestKey: string]: SubscriptionData
 }
 
 export const subscriptionsReducer = createReducer<SubscriptionState>({}, (builder) => {
-  builder.addCase(actions.warmupSubscribed, (state, action) => {
-    state[getSubscriptionKey(action.payload)] = {
-      origin: action.payload,
-      executeFn: action.payload.executeFn,
-      startedAt: Date.now(),
-      isDuplicate: !!state[getSubscriptionKey(action.payload)],
+  builder.addCase(actions.warmupSubscribed, (state, { payload }) => {
+    const key = payload.key || getSubscriptionKey(payload)
+
+    state[key] = {
+      origin: payload.data,
+      executeFn: payload.executeFn,
+      startedAt: state[key]?.startedAt ?? Date.now(),
+      isDuplicate: !!state[key],
+      parent: payload.parent || state[key]?.parent,
+      batchablePropertyPath: payload.batchablePropertyPath || state[key]?.batchablePropertyPath,
+      childLastSeenById: payload?.childLastSeenById,
     }
   })
 
   builder.addCase(actions.warmupUnsubscribed, (state, action) => {
+    const subscription = state[action.payload.key]
+
+    if (!subscription)
+      console.error(
+        'subscriptionsReducer warmupUnsubscribed - no subscription found',
+        JSON.stringify(state),
+      )
+
+    if (subscription.childLastSeenById) {
+      const children = Object.keys(subscription.childLastSeenById)
+      for (const childKey of children) {
+        delete state[childKey]
+      }
+    }
     delete state[action.payload.key]
   })
+
+  builder.addCase(actions.warmupJoinGroup, (state, { payload }) => {
+    if (!payload)
+      console.error('subscriptionsReducer warmupJoinGroup - no payload', JSON.stringify(state))
+
+    const batchWarmer = state[payload.parent]
+    if (!batchWarmer) {
+      console.error(
+        'subscriptionsReducer warmupJoinGroup - no batch warmer exists',
+        JSON.stringify(state),
+      )
+    }
+
+    batchWarmer.childLastSeenById = {
+      ...batchWarmer.childLastSeenById,
+      ...payload.childLastSeenById,
+    }
+    for (const childKey in payload.childLastSeenById) {
+      const childRequestData = state[childKey]?.origin
+      if (childRequestData) {
+        for (const path of payload.batchablePropertyPath) {
+          const uniqueBatchableValue = new Set(batchWarmer.origin[path])
+          uniqueBatchableValue.add(childRequestData[path])
+          batchWarmer.origin[path] = [...uniqueBatchableValue]
+        }
+      }
+    }
+  })
+
+  builder.addCase(actions.warmupLeaveGroup, (state, { payload }) => {
+    if (!payload)
+      console.error('subscriptionsReducer warmupLeaveGroup - no payload', JSON.stringify(state))
+
+    const batchWarmer = state[payload.parent]
+    if (!batchWarmer) {
+      console.error(
```

Comment:
We should be using the logger everywhere instead of `console`

Reply:
Oops, meant to do that. Changing.